### PR TITLE
[Issue 8260] Support reset cursor to a batch index of the batching message

### LINKED ISSUE #8260: Support reset cursor to a batch index of the batching message
**Is your feature request related to a problem? Please describe.**
We have support ack the batch message index since 2.6.0. But it's not supported by the reset cursor operation. So, if users reset the cursor to a batch index of the batch message then when consuming messages again, it will get these batch index again.

### PR DESCRIPTION
**Pull request by Renkai**
Signed-off-by: Renkai <gaelookair@gmail.com>

*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #8260

### Motivation

Make reset cursor command able to reset to a specific index in batch mode.

### Modifications

Now reset offset command change index info on broker side

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (yes / no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by Renkai** [comment]
Still have some problem, unit test written by me can't be passed(due to timeout), 
https://github.com/apache/pulsar/pull/8285/files#diff-30ab326ad2d2e0d984b321af3c68c7bb7ab345fac245bbcc6739f58daffc0e68R120

Maybe there is some thing wrong with my test configuration
```
mvn -pl pulsar-broker -Dtest=SubscriptionSeekTest\#testSeekForBatch test -DredirectTestOutputToFile=false
```

PTAL @codelipenghui

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 1]
```diff
@@ -36,10 +36,7 @@
 import io.netty.handler.ssl.SslHandler;
 
 import java.net.SocketAddress;
-import java.util.List;
-import java.util.Map;
-import java.util.NoSuchElementException;
-import java.util.Set;
+import java.util.*;
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:39]: Avoid use import .* here.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/SubscriptionSeekTest.java [hunk 2]
```diff
@@ -25,20 +25,21 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 
 import lombok.extern.slf4j.Slf4j;
 import org.apache.pulsar.broker.service.persistent.PersistentSubscription;
 import org.apache.pulsar.broker.service.persistent.PersistentTopic;
 import org.apache.pulsar.client.admin.PulsarAdminException;
-import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.*;
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/SubscriptionSeekTest.java:41]: Avoid use import .* here.

File: pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java [hunk 3]
```diff
@@ -770,14 +770,15 @@ public static ByteBuf newActiveConsumerChange(long consumerId, boolean isActive)
...
-    public static ByteBuf newSeek(long consumerId, long requestId, long ledgerId, long entryId) {
+    public static ByteBuf newSeek(long consumerId, long requestId, long ledgerId, long entryId, int batchIndex) {
         CommandSeek.Builder seekBuilder = CommandSeek.newBuilder();
         seekBuilder.setConsumerId(consumerId);
         seekBuilder.setRequestId(requestId);
 
         MessageIdData.Builder messageIdBuilder = MessageIdData.newBuilder();
         messageIdBuilder.setLedgerId(ledgerId);
         messageIdBuilder.setEntryId(entryId);
+        messageIdBuilder.setBatchIndex(batchIndex);
```
Comment (codelipenghui) [inline pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java:781]: The `batchIndex` of the MessageIdData is an optional field, we shouldn't always use '0' here. It better to change the params type of the `batchIndex` to `Integer`. If the the `batchIndex` no a null value, set the value for the `MessageIdBuilder`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 4]
```diff
@@ -1349,8 +1347,14 @@ protected void handleSeek(CommandSeek seek) {
...
             MessageIdData msgIdData = seek.getMessageId();
+            BitSetRecyclable ackSet = BitSetRecyclable.create();
+            if (msgIdData.hasBatchIndex()) {
+                ackSet.set(0, Math.max(msgIdData.getBatchIndex(), 0));
+            }
 
-            Position position = new PositionImpl(msgIdData.getLedgerId(), msgIdData.getEntryId());
+            Position position = new PositionImpl(msgIdData.getLedgerId(),
+                    msgIdData.getEntryId(), ackSet.toLongArray());
+            ackSet.recycle();
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:1357]: It's better tot check `msgIdData.hasBatchIndex()` first, If the conditions are not met, we don't need to create the BitSet instance.

**Review by codelipenghui — CHANGES_REQUESTED** [review]
I have double-checked the implementation of the `Batch Index Acknowledgment`. Looks we can only pass the `ackset` to the broker from the client-side, because currently `true` in the `BitSet` represent the message does not acked, and `false` means the batch index are acked. So I think it's the root cause of why the consumer start receive message from the next message of the reset position.

**Comment by wolfstudy** [comment]
ping @codelipenghui PTAL again thanks.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/SubscriptionSeekTest.java [hunk 5]
```diff
@@ -116,6 +122,91 @@ public void testSeek() throws Exception {
...
+
+        futureMessageIds.forEach(future -> {
+            MessageId messageId = null;
+            try {
+                messageId = future.get();
+                messageIds.add(messageId);
+            } catch (InterruptedException | ExecutionException e) {
+                e.printStackTrace();
+            }
+        });
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/SubscriptionSeekTest.java:155]:
```suggestion
        FutureUtil.waitForAll(futureMessageIds).get()
```

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/SubscriptionSeekTest.java [hunk 6]
```diff
@@ -116,6 +122,91 @@ public void testSeek() throws Exception {
...
+        producer.flush();
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/SubscriptionSeekTest.java:156]: Since the messageId is returned, we don't need to flush here.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/SubscriptionSeekTest.java [hunk 7]
```diff
@@ -116,6 +122,91 @@ public void testSeek() throws Exception {
...
+        consumer.seek(resetId);
+        // Wait for consumer to reconnect
+        Thread.sleep(500);
+
+        nextMessage = consumer.receive();
+        nextId = nextMessage.getMessageId();
+        consumer.acknowledge(nextId);
+        expectedMessage = messages.get(2);
+        log.info("expected next message3: {}, next message3 {}", expectedMessage, nextMessage.getValue());
+        assertEquals(nextMessage.getValue(), expectedMessage);
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/SubscriptionSeekTest.java:206]: Could you please also add some corner case test? such as reset to the last message to check the consumer can't receive any message?

**Review by codelipenghui — APPROVED** [review]

**Comment by codelipenghui** [comment]
cherry-picked to branch-2.6(2.6.2)
